File: packages/core/types/email-hub.ts
import { z } from "zod";

export const EmailHubAccountSchema = z.object({
  id: z.string(),
  email_address: z.string(),
  provider: z.string(),
  connected_at: z.string(),
  last_sync_at: z.string().optional(),
});

export const EmailHubWatchSchema = z.object({
  changed: z.boolean(),
  synced: z.boolean(),
  at: z.string().optional(),
});

export const EmailHubInboxWatchSchema = z.object({
  subscribed: z.boolean(),
});

export const EmailHubSyncSchema = z.object({
  synced: z.boolean(),
});

export const EmailHubUnreadSchema = z.object({
  unread: z.number().optional().default(0),
});

export const EmailHubCountsSchema = z.object({
  total: z.number(),
  unread: z.number(),
});

export const EmailHubAttachmentSchema = z.object({
  id: z.string(),
  filename: z.string(),
  mime_type: z.string(),
  size_bytes: z.number(),
});

export const EmailHubThreadSchema = z.object({
  id: z.string(),
  account_id: z.string(),
  folder: z.string(),
  subject: z.string(),
  snippet: z.string(),
  from_addr: z.string(),
  from_name: z.string().optional(),
  to_addrs: z.array(z.string()),
  sent_at: z.string(),
  is_read: z.boolean(),
  is_starred: z.boolean(),
  has_attachments: z.boolean(),
  attachments: z.array(EmailHubAttachmentSchema).optional(),
  body_text: z.string().optional(),
  body_html: z.string().optional(),
  body_cached: z.boolean(),
  imap_labels: z.array(z.string()).optional().default([]),
  snoozed_until: z.string().optional(),
});

export const EmailHubImapLabelListSchema = z.object({
  labels: z.array(z.string()).optional().default([]),
});

export const EmailHubAccountListSchema = z.object({
  accounts: z.array(EmailHubAccountSchema),
});

export const EmailHubThreadListSchema = z.object({
  threads: z.array(EmailHubThreadSchema),
  counts: EmailHubCountsSchema,
  next_cursor: z.string().optional(),
});

export const EmailHubSummaryActionItemSchema = z.object({
  title: z.string(),
  owner: z.string().optional().default(""),
  due: z.string().optional().default(""),
});

export const EmailHubThreadSummarySchema = z.object({
  summary: z.string(),
  key_points: z.array(z.string()).optional().default([]),
  action_items: z.array(EmailHubSummaryActionItemSchema).optional().default([]),
  needs_reply: z.boolean().optional().default(false),
  reply_hint: z.string().optional().default(""),
  model: z.string().optional().default(""),
  cached: z.boolean().optional().default(false),
  summarized_at: z.string().optional().default(""),
});

export type EmailHubAttachment = z.infer<typeof EmailHubAttachmentSchema>;
export type EmailHubThreadSummary = z.infer<typeof EmailHubThreadSummarySchema>;
export type EmailHubSummaryActionItem = z.infer<typeof EmailHubSummaryActionItemSchema>;

export interface EmailHubThreadFilters {
  q?: string;
  from?: string;
  label?: string;
  unreadOnly?: boolean;
  hasAttachmentsOnly?: boolean;
}

export type EmailHubAccount = z.infer<typeof EmailHubAccountSchema>;
export type EmailHubWatch = z.infer<typeof EmailHubWatchSchema>;
export type EmailHubInboxWatch = z.infer<typeof EmailHubInboxWatchSchema>;
export type EmailHubSync = z.infer<typeof EmailHubSyncSchema>;
export type EmailHubThread = z.infer<typeof EmailHubThreadSchema>;
export type EmailHubCounts = z.infer<typeof EmailHubCountsSchema>;

File: server/internal/handler/dto/sdo/email_hub.go
package sdo

// EmailHubAccountSDO is one connected IMAP mailbox.
type EmailHubAccountSDO struct {
	ID           string `json:"id" example:"01JABC1234567890ABCDEFGH"`
	EmailAddress string `json:"email_address" example:"you@gmail.com"`
	Provider     string `json:"provider" example:"gmail"`
	ConnectedAt  string `json:"connected_at" format:"date-time"`
	LastSyncAt   string `json:"last_sync_at,omitempty" format:"date-time"`
}

// EmailHubWatchSDO is POST .../email-hub/watch.
type EmailHubWatchSDO struct {
	Changed bool   `json:"changed" example:"true"`
	Synced  bool   `json:"synced" example:"true"`
	At      string `json:"at,omitempty" format:"date-time"`
}

// EmailHubInboxWatchSDO is POST/DELETE .../email-hub/inbox-watch.
type EmailHubInboxWatchSDO struct {
	Subscribed bool `json:"subscribed" example:"true"`
}

// EmailHubSyncSDO is POST .../email-hub/sync.
type EmailHubSyncSDO struct {
	Synced bool `json:"synced" example:"true"`
}

// EmailHubUnreadSDO is GET .../email-hub/unread-count.
type EmailHubUnreadSDO struct {
	Unread int64 `json:"unread" example:"3"`
}

// EmailHubScheduledSendItemSDO is one open scheduled outbound message: pending
// (waiting to go out) or failed (delivery gave up; retry or dismiss). The
// failure reason is not exposed.
type EmailHubScheduledSendItemSDO struct {
	ID      string   `json:"id"`
	SendAt  string   `json:"send_at" format:"date-time"`
	Subject string   `json:"subject"`
	To      []string `json:"to"`
	Status  string   `json:"status" enum:"pending,failed" example:"pending"`
}

// EmailHubScheduledSendListSDO lists the caller's open scheduled sends,
// failed ones first, then by send_at ascending.
type EmailHubScheduledSendListSDO struct {
	Scheduled []EmailHubScheduledSendItemSDO `json:"scheduled"`
}

// EmailHubScheduledSendSDO is POST .../email-hub/send when send_at is in the future.
type EmailHubScheduledSendSDO struct {
	Scheduled bool     `json:"scheduled" example:"true"`
	ID        string   `json:"id" example:"01JABC1234567890ABCDEFGH"`
	SendAt    string   `json:"send_at" format:"date-time"`
	Subject   string   `json:"subject" example:"Hello from UniWork"`
	To        []string `json:"to" example:"[\"client@example.com\"]"`
	AccountID string   `json:"account_id" example:"01JABC1234567890ABCDEFGH"`
}

// EmailHubAccountListSDO lists connected mailboxes for the caller.
type EmailHubAccountListSDO struct {
	Accounts []EmailHubAccountSDO `json:"accounts"`
}

// EmailHubAttachmentSDO is cached attachment metadata for one thread row.
type EmailHubAttachmentSDO struct {
	ID        string `json:"id"`
	Filename  string `json:"filename"`
	MimeType  string `json:"mime_type"`
	SizeBytes int64  `json:"size_bytes"`
}

// EmailHubThreadSDO is a cached inbox row.
type EmailHubThreadSDO struct {
	ID                       string                  `json:"id"`
	AccountID                string                  `json:"account_id"`
	Folder                   string                  `json:"folder"`
	Subject                  string                  `json:"subject"`
	Snippet                  string                  `json:"snippet"`
	FromAddr                 string                  `json:"from_addr"`
	FromName                 string                  `json:"from_name,omitempty"`
	ToAddrs                  []string                `json:"to_addrs"`
	SentAt                   string                  `json:"sent_at" format:"date-time"`
	IsRead                   bool                    `json:"is_read"`
	IsStarred                bool                    `json:"is_starred"`
	HasAttachments           bool                    `json:"has_attachments"`
	Attachments              []EmailHubAttachmentSDO `json:"attachments,omitempty"`
	BodyText                 string                  `json:"body_text,omitempty"`
	BodyHTML                 string                  `json:"body_html,omitempty"`
	BodyCached               bool                    `json:"body_cached"`
	ImapLabels               []string                `json:"imap_labels,omitempty"`
	SnoozedUntil             string                  `json:"snoozed_until,omitempty" format:"date-time"`
	ConversationMessageCount int                     `json:"conversation_message_count,omitempty"`
}

// EmailHubImapLabelListSDO lists distinct user-visible IMAP labels on cached threads.
type EmailHubImapLabelListSDO struct {
	Labels []string `json:"labels"`
}

// EmailHubThreadListSDO is a folder listing with counts.
type EmailHubThreadListSDO struct {
	Threads    []EmailHubThreadSDO `json:"threads"`
	Counts     EmailHubCountsSDO   `json:"counts"`
	NextCursor string              `json:"next_cursor,omitempty"`
}

// EmailHubConversationSDO lists every message in one conversation (oldest first).
type EmailHubConversationSDO struct {
	Messages []EmailHubThreadSDO `json:"messages"`
}

// EmailHubCountsSDO holds folder totals.
type EmailHubCountsSDO struct {
	Total  int64 `json:"total"`
	Unread int64 `json:"unread"`
}

// EmailHubThreadSummarySDO is POST .../email-hub/threads/{threadID}/ai/summarize.
type EmailHubThreadSummarySDO struct {
	Summary      string                 `json:"summary"`
	KeyPoints    []string               `json:"key_points"`
	ActionItems  []SummaryActionItemDTO `json:"action_items"`
	NeedsReply   bool                   `json:"needs_reply"`
	ReplyHint    string                 `json:"reply_hint,omitempty"`
	Model        string                 `json:"model,omitempty"`
	Cached       bool                   `json:"cached"`
	SummarizedAt string                 `json:"summarized_at,omitempty" format:"date-time"`
}
